src/Engine/ProtoCore/Lang/FunctionGroup.cs

```diff
@@ -307,6 +335,86 @@ public void CopyPublic(List<FunctionEndPoint> rhs)
             return ret;
         }
 
+        internal static bool CheckInvalidArrayCoersion(CLRFunctionEndPoint fep, List<CLRStackValue> reducedSVs, MSILRuntimeCore runtimeCore, bool allowArrayPromotion)
+        {
+            ClassTable classTable = runtimeCore.ClassTable;
+
+            for (int i = 0; i < reducedSVs.Count; i++)
+            {
+                Type typ = fep.FormalParams[i].ProtoInfo;
+                if (typ.UID == (int)ProtoCore.PrimitiveType.InvalidType)
+                    return true;
+
+                if (!typ.IsIndexable)
+                    continue; //It wasn't an array param, skip
+
+                //Compute the type of target param
+                if (!allowArrayPromotion)
+                    Validity.Assert(reducedSVs[i].IsEnumerable, "This should be an array otherwise this shouldn't have passed previous tests");
+
+                if (!allowArrayPromotion)
+                {
+                    if (typ.rank != ArrayUtils.GetMaxRankForArray(reducedSVs[i]) &&
+                        typ.rank != DSASM.Constants.kArbitraryRank)
+                        return true; //Invalid co-ercsion
+                }
+                else
+                {
+                    if (typ.rank < ArrayUtils.GetMaxRankForArray(reducedSVs[i]) &&
+                        typ.rank != DSASM.Constants.kArbitraryRank)
+                        return true; //Invalid co-ercsion
+
+                }
+
+                Dictionary<ClassNode, int> arrayTypes = ArrayUtils.GetTypeStatisticsForArray(reducedSVs[i], runtimeCore);
+
+                ClassNode cn = null;
+
+                if (arrayTypes.Count == 0)
+                {
+                    //This was an empty array
+                    Validity.Assert(cn == null, "If it was an empty array, there shouldn't be a type node");
+                    cn = runtimeCore.ClassTable.ClassNodes[(int)PrimitiveType.Null];
+                }
+                else if (arrayTypes.Count == 1)
+                {
+                    //UGLY, get the key out of the array types, of which there is only one
+                    foreach (ClassNode key in arrayTypes.Keys)
```

Comment:
then why use a foreach?

Reply:
😆

Reply:
@pinzart90 looks like you fixed this in the copied function, but not here:
`cn = arrayTypes.Keys.First();`